### PR TITLE
add docs re generic assembler for butler composite

### PR DESCRIPTION
**Pull request by n8pease**

### CODE REVIEW

File: index.rst [hunk 1]
```diff
@@ -632,15 +641,76 @@ The disassembler function signature is:
...
+
+ComponentInfo is a class used when assembling and disassembling a composite
+object in butler. It contains information about a component of a composite
+object and has a slot for passing the component object into the assembler and for
+the disssembler to pass the component object out.
+
+ComponentInfo is used as an input to assemblers and disassemblers (which are
+part of the butler public API).
+
+Some ComponentInfo parameters are populated with information from the policy and
```
Comment (brianv0) [inline index.rst:677]: Just a question: Are the policy and the butler the only things which can fill in ComponentInfo parameters? If that's the case, maybe reword this stanza? Also, is the first part referring to assembling only?

> Some ComponentInfo parameters are populated with information from the policy, and the others are populated by the butler. When disassembling a composite object, the obj slot is populated by the disassembler. More details are available in the class docstrings.

File: index.rst [hunk 2]
```diff
@@ -632,15 +641,76 @@ The disassembler function signature is:
...
         dataId : dict
             The dataId that is being used to persist this object.
-        componentDict : dict
-            A dict to populate with the components that should be persisted for
-            this composite.
+        componentInfo : dict of componentDatasetName to ComponentInfo
+            A dict of ComponentInfo instances to populate with the components
+            that should be persisted for this composite. Keys are the component
+            names as defined by the datasetType in the policy. Values are
+            ComponentInfo objects.
```
Comment (ktlim) [inline index.rst:659]: Should mention that only `ComponentInfo.obj` is to be modified; the rest must be left alone.

File: index.rst [hunk 3]
```diff
@@ -632,15 +641,76 @@ The disassembler function signature is:
...
         None
         """
 
+ComponentInfo
+"""""""""""""
+
+ComponentInfo is a class used when assembling and disassembling a composite
+object in butler. It contains information about a component of a composite
+object and has a slot for passing the component object into the assembler and for
+the disssembler to pass the component object out.
```
Comment (ktlim) [inline index.rst:672]: Typo: "disassembler"

File: index.rst (hunk 1 above)
Comment (ktlim) [reply index.rst:677]: Do you anticipate that the setter and getter may be useful for more than the generic assembler/disassembler?  If not, it might be nice to pass them separately to that function (i.e. the generic assembler/disassembler would have a different signature than a policy-specified one) which would allow simplification of the `ComponentInfo` class (perhaps even back to eliminating it altogether).

File: index.rst [hunk 4]
```diff
@@ -632,15 +641,76 @@ The disassembler function signature is:
...
+""""""""""""""""""""""""""""""""
+
+Butler has a generic assembler and disassembler that can be used in some cases,
+listed below. The list is in the order butler will try to apply the generic
+assembler.
+
+The generic assembler will be used when:
+
+1. The policy for the composite does not name an assembler.
+2. If a setter is not named for any component of the composite, and the python
```
Comment (ktlim) [inline index.rst:692]: Not named _in the policy_ (as opposed to in the code).

File: index.rst [hunk 5]
```diff
@@ -632,15 +641,76 @@ The disassembler function signature is:
...
+Some ComponentInfo parameters are populated with information from the policy and
+some are filled in by the butler, and when disassembling a composite object the
+obj slot is populated by the disassembler. More details are available in the
+class docstrings.
+
+Generic Assembler & Disassembler
+""""""""""""""""""""""""""""""""
+
+Butler has a generic assembler and disassembler that can be used in some cases,
+listed below. The list is in the order butler will try to apply the generic
```
Comment (ktlim) [inline index.rst:686]: The generic assembler is tried whenever one is not named in the policy (item 1).  It _works_ when any of item 2 or item 3 hold.  It is _not_ required that all of items 1, 2, and 3 simultaneously hold.

**Review by ktlim — APPROVED** [review]
